File: binary_tree_aximum_path_sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int res = root->val;
        dfs(root, res);
        return res;
    }

private:
    int dfs(TreeNode* node, int& res) {
        if (!node) {
            return 0;
        }

        // Recursively compute the maximum sum of the left and right subtree paths.
        int leftSum = max(0, dfs(node->left, res));
        int rightSum = max(0, dfs(node->right, res));

        // Update the maximum path sum encountered so far (with split).
        res = max(res, leftSum + rightSum + node->val);

        // Return the maximum sum of the path (without split).
        return max(leftSum, rightSum) + node->val;
    }
};

File: binary_tree__aximum_path_sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = root.val

        def dfs(node):
            nonlocal res

            if not node:
                return 0
            
            # recursively compute the maximum sum of the left and right subtree paths
            left_sum = max(0, dfs(node.left))
            right_sum = max(0, dfs(node.right))

            # update the maximum path sum encountered so far(with split)
            res = max(res, left_sum + right_sum + node.val)

            # return the maximum sum of the path(without split)
            return max(left_sum, right_sum) + node.val

        dfs(root)
        return res
